packages/modules/common/abstract_component.py

```diff
@@ -1,106 +1,60 @@
 from abc import abstractmethod
 from typing import List, Union
 try:
-    from ...helpermodules import log
-    from ..common import modbus
-    from ..common import lovato
-    from ..common import mpm3pm
-    from ..common import simcount
-    from ..common import sdm630
-    from ..common import store
     from ..common.module_error import ModuleError, ModuleErrorLevel
     from component_state import BatState, CounterState, InverterState
 except:
-    from helpermodules import log
-    from modules.common import modbus
-    from modules.common import lovato
-    from modules.common import mpm3pm
-    from modules.common import simcount
-    from modules.common import sdm630
-    from modules.common import store
     from modules.common.module_error import ModuleError, ModuleErrorLevel
     from .component_state import BatState, CounterState, InverterState
 
 
-class AbstractComponent:
-    def __init__(self, device_id: int, component_config: dict, client: Union[modbus.ModbusClient, mpm3pm.Mpm3pm, lovato.Lovato, sdm630.Sdm630]) -> None:
-        try:
-            self.device_id = device_id
-            self.client = client
-            self.data = {"config": component_config,
-                         "simulation": {}}
-            self.value_store = (store.ValueStoreFactory().get_storage(component_config["type"]))(self.data["config"]["id"])
-            simcount_factory = simcount.SimCountFactory().get_sim_counter()
-            self.sim_count = simcount_factory()
-        except Exception as e:
-            log.MainLogger().exception("Fehler im Modul "+self.data["config"]["name"])
+class AbstractCounter:
+    @abstractmethod
+    def __init__(self, device_id: int, component_config: dict, tcp_client, sim_count) -> None:
+        pass
 
     @abstractmethod
-    def get_values(self) -> Union[BatState, CounterState, InverterState]:
+    def get_values(self) -> CounterState:
         pass
 
+
+class CounterUpdater:
+    class State:
+        def __init__(self, voltages: List[float], currents: List[float], powers: List[float], power_factors: List[float], imported: float, exported: float, power_all: float, frequency: float):
+            self.voltages = voltages
+            self.currents = currents
+            self.powers = powers
+            self.power_factors = power_factors
+            self.imported = imported
+            self.exported = exported
+            self.power_all = power_all
+            self.frequency = frequency
+
+    def __init__(self, counter, value_store):
+        self.value_store = value_store
+        self.counter = counter
+
+
+class ComponentUpdater:
+    def __init__(self, component: CounterUpdater) -> None:
+        self.component = component
+
     def update_values(self) -> Union[BatState, CounterState, InverterState]:
```

Comment:
Das passt noch nicht so richtig zusammen. Also die Klasse heißt `ComponentUpdater`. Der Konstruktor bekommt einen Parameter `component: CounterUpdater`. Weiter unten kommt noch `component.counter.get_values()` und `component.value_store.set(state)`.

Also ich glaube wir müssen uns erstmal über die Begriffe einig werden. Mein Verständnis ist bisher, dass es in 2.0 so definiert ist:
1. Ein **device** ist ein pysikalisches Gerät wie zum Beispiel ein Wechselrichter
2. Eine **component** ist ein Baustein dieses Geräts. Zum Beispiel hat ein Wechselrichter neben einer Leistung auch noch einen EVU-Zähler eingebaut. So ein WR hat dann zwei "components": Nämlich die eigene Leistung und die Daten vom EVU-Zähler.

Ist das richtig? Wenn ja, dann nächster Punkt: Es gibt (bisher) drei Arten von **Component**: **Inverter**, **Counter** und **Battery**. Richtig?

Dann gibt es den **ValueStore**. Das ist quasi die "Datenbank". Jeder ValueStore verarbeitet einen Datentyp. Daher haben wir drei verschiedene *ValueStore*: Einen für jede Art von *Component*.

Wenn ich das jetzt nicht richtig habe, dann ist alles was jetzt noch kommt an Text Käse.

Nach der Logik stelle ich mir vor, dass der **ComponentUpdater** die Werte von einer *Component* abruft und den erhaltenen Wert dem **ValueStore** übergibt. Fein, das sieht auch ungefähr so aus, als bekomme ich das in der Methode `update_values`. Aber ein paar Dinge passen nicht:
1. Warum gibt es im Konstruktor einen Parameter `component: CounterUpdater`. Wenn ich den Aufbau richtig verstanden habe, stimmt entweder der Name nicht oder der Typ nicht.
2. Warum gibt `update_values` ein `Union[BatState, CounterState, InverterState]` zurück? Also eigentlich würde ich sagen, dass muss garnichts zurückgeben, aber wenn man es denn braucht, dann wäre es gut, wenn der `ComponentUpdater` entweder spezifisch ist und sich um genau einen Typ kümmert (der dann auch bekannt ist) oder er generisch ist und er garnicht genau weiß, was für Daten er rumschubst. Letzteres ließe sich sehr schön mit einem Generic lösen.
3. Warum hat die Klasse `CounterUpdater` keine Methode? Wenn die Klasse `...Updater` heißt, dann muss die was tun können. Also zum Beispiel was updaten können. Da müsste zumindest eine Methode `update()` oder sowas dran sein. Ist es aber nicht. `CounterUpdater` ist einfach nur ein Verbund aus `counter` und einem `value_store`.

Es ist verständlich, dass so eine generische Aufgabe, wie der `ComponentUpdater` nicht drei mal implementiert werden soll. Das würde tatsächlich auch drei mal das gleiche machen. Lösen lässt sich das mit einem `Generic`. Zum Beispiel so:
```python
T = TypeVar('T')


class ValueStore(Generic[T]):
    @abstractmethod
    def set(self, value: T) -> None:
        pass


class Component(Generic[T]):
    @abstractmethod
    def get_state(self) -> T:
        pass


class ComponentUpdater(Generic[T]):
    def __init__(self, component: Component[T], value_store: ValueStore[T]) -> None:
        self.component = component
        self.value_store = value_store

    def update_values(self) -> None:
        self.value_store.set(self.component.get_state())
```
Der Clue ist: `T` fungiert hier als ein Platzhalter. In der Praxis ist T entweder `BatState`, `CounterState` oder `InverterState`. Der Knackpunkt ist vorallem ` component: Component[T], value_store: ValueStore[T]` (in der `__init__`): Hier wird deutlich gemacht: Es ist zwar unerheblich welchen Typ die `Component` hat und welcher Typ der `ValueStore`, aber beide müssen zwangsweise den *gleichen* Typ haben.
```

Über den Rest können wir nochmal schreiben, nachdem wir uns auf die Terminologie geeinigt haben.

Reply:
Das ist richtig zusammengefasst.

Zu 1.: Der Typ muss auch generisch sein. 
Zu 2.: Der Rückgabewert wird für manche Module, wie z.B. Solaredge benötigt, da die Werte der Komponenten noch miteinander verrechnet werden müssen. Dafür muss auch noch das Lesen und Schreiben der Werte getrennt werden.